### PR TITLE
update the metric RFC

### PR DESCRIPTION
**Pull request by superp0tat0**
Update the metric RFC based on the meeting at June 21

### CODE REVIEW

File: attic/rfc_metrics.md [hunk 1]
```diff
@@ -36,14 +42,24 @@ Internally, the function would look the following:
...
+  # minimize, properties, ...
+  self$minimize = measure$minimize
 }
+...
 
-fn_cm = function(m, na_value = NaN) {
-  m[2L, 1L]
-}
+$.score = function(task, operation, ...) {
+m1 = self$measure$score(group1, )
```
Comment (pfistfl) [inline attic/rfc_metrics.md:56]: How do you get the groups? What helper functions are needed?

File: attic/rfc_metrics.md [hunk 2]
```diff
@@ -36,14 +42,24 @@ Internally, the function would look the following:
...
+  # here do some other thing setting task types etc.
+  # minimize, properties, ...
+  self$minimize = measure$minimize
 }
+...
 
-fn_cm = function(m, na_value = NaN) {
-  m[2L, 1L]
-}
+$.score = function(task, operation, ...) {
```
Comment (pfistfl) [inline attic/rfc_metrics.md:55]: I think `operation` should not be passed during `$score` but during initialization. 
It should default to abs_diff

File: attic/rfc_metrics.md [hunk 3]
```diff
@@ -36,14 +42,24 @@ Internally, the function would look the following:
...
+m1 = self$measure$score(group1, )
+m2 = self$measure$score....
```
Comment (pfistfl) [inline attic/rfc_metrics.md:57]: Try to fill in the ... so we can see possible missing data etc.

File: attic/rfc_metrics.md [hunk 4]
```diff
@@ -36,14 +42,24 @@ Internally, the function would look the following:
...
 Example:
 ```r
-fn = function(truth, response, positive, ...) {
-  assert_binary(truth, response = response, positive = positive)
-  fn_cm(cm(truth, response, positive))
+MeasureFairnessSimple = R6Class(
+inherits = "Measure",
+initialize = function(measure) {
+  self$measure = measure
+  # here do some other thing setting task types etc.
```
Comment (pfistfl) [inline attic/rfc_metrics.md:49]: what are the other properties. Try to find that out

File: attic/rfc_metrics.md (hunk 2 above)
Comment (superp0tat0) [reply attic/rfc_metrics.md:55]: I see, it should be set when creating the measure

Comment (superp0tat0) [reply attic/rfc_metrics.md:55]: Emm, got one question. If we are gonna set the operation when initialize the fairness measures. Like
```
fair_measure = msr("groupwise_abs_diff", measure = base_measure)
```
Then there is no need to pass in the operation.

File: attic/rfc_metrics.md (hunk 1 above)
Comment (pfistfl) [reply attic/rfc_metrics.md:56]: If its simple just abstract, otherwise include it

File: attic/rfc_metrics.md (hunk 2 above)
Comment (pfistfl) [reply attic/rfc_metrics.md:55]: I agree. 
Let us ignore the operation aspect for now and focus on getting absdiff to work.

File: attic/rfc_metrics.md [hunk 5]
```diff
@@ -36,14 +46,55 @@ Internally, the function would look the following:
...
+    #' Creates a new instance of this [R6][R6::R6Class] class.
+    initialize = function(name, measure) {
+      info = mlr3fairness::measures[[name]]
+      super$initialize(
+        id = paste0("fairness.", name),
+        range = c(info$lower, info$upper),
+        minimize = info$minimize,
+        predict_type = info$predict_type,
+        packages = "mlr3fairness",
+        man = paste0("mlr3::mlr_measures_fairness.", name)
```
Comment (pfistfl) [inline attic/rfc_metrics.md:64]: I think the `man` (this is where to find documentation does not make sense. 
I would set it to 'mlr_measures_fairness`

Also: Are you sure you need `mlr3::`?

File: attic/rfc_metrics.md [hunk 6]
```diff
@@ -36,14 +46,55 @@ Internally, the function would look the following:
...
-fn_cm = function(m, na_value = NaN) {
-  m[2L, 1L]
+The groupwise operations are performed inside the Fairness Metric functions. See below:
+```r
+groupwise_abs_diff <- function(prediction, base_measure, positive, data_task, response = NULL, ...){
+  #Assert the status for all the parameters
+
+  subcol = data_task$col_roles$pta
+  prediction = split(as.data.table(prediction), data_task$data(cols = subcol))
+  prediction = lapply(prediction, as_prediction_classif)
```
Comment (pfistfl) [inline attic/rfc_metrics.md:89]: we use `map` from `mlr3misc` here instead of `lapply`.

Comment (pfistfl) [reply attic/rfc_metrics.md:89]: Also: Does `Prediction` have a `filter`  or `subset `method we can use?

File: attic/rfc_metrics.md [hunk 7]
```diff
@@ -36,14 +46,55 @@ Internally, the function would look the following:
...
-  fn_cm(cm(truth, response, positive))
-}
+MeasureFairness = R6Class("MeasureFairness", inherit = Measure, cloneable = FALSE,
+  public = list(
+    fun = NULL,
+    na_value = NaN,
+    base_measure = NULL,
+    #' @description
+    #' Creates a new instance of this [R6][R6::R6Class] class.
+    initialize = function(name, measure) {
```
Comment (pfistfl) [inline attic/rfc_metrics.md:56]: let us get rid of `name` here for now and always do `groupwise_abs_diff`

File: attic/rfc_metrics.md [hunk 8]
```diff
@@ -36,14 +46,55 @@ Internally, the function would look the following:
...
+      info = mlr3fairness::measures[[name]]
+      super$initialize(
+        id = paste0("fairness.", name),
```
Comment (pfistfl) [inline attic/rfc_metrics.md:59]: This should really be `measure.name` so we know which measure is considered.

Comment (pfistfl) [reply attic/rfc_metrics.md:59]: or info$name or whatever:)

File: attic/rfc_metrics.md [hunk 9]
```diff
@@ -20,13 +20,23 @@ I propose when applying a metric method, what occurs should be ...
...
 Example:
 ```r
-#The true label of the observations
-truth = as.factor(c(1,1,1,0,0,1,1,0,0,0))
-
-#The predicted labels of the observations
-response = as.factor(c(1,1,0,0,1,1,0,0,0,1))
-
-fn(truth, response, positive = "1")
+#Initialize the data task and learner task
```
Comment (pfistfl) [inline attic/rfc_metrics.md:23]: Feel free to add a few more sentences to the RFC.

Comment (pfistfl) [reply attic/rfc_metrics.md:23]: Describe in detail what the current open questions are.
This is the basis for discussions with the other Mentors

File: attic/rfc_metrics.md (hunk 5 above)
Comment (superp0tat0) [reply attic/rfc_metrics.md:64]: Emm, should we set to mlr3fairness:: or just ignore it.

File: attic/rfc_metrics.md (hunk 6 above)
Comment (superp0tat0) [reply attic/rfc_metrics.md:89]: Have changed lapply to map.
There is no such functions in prediction, cause prediction does not contain the data needed for predictions. Which is also the reason why need to input the data task at ```prediction$score(fairness_measure, data_task)```

**Pull request merged**